Userland/Applications/Maps/MapsWidget.cpp

```diff
@@ -0,0 +1,182 @@
+    , m_attribution_text(options.attribution_text)
+    , m_attribution_url(options.attribution_url)
+{
+}
+
+void MapsWidget::mousedown_event(GUI::MouseEvent& event)
+{
+    if (event.buttons() != GUI::MouseButton::Primary)
+        return;
+
+    // Handle attribution text click
+    if (m_attribution_enabled) {
+        GUI::Painter painter(*this);
+        float attribution_width = 8 + painter.font().width(m_attribution_text) + 8;
+        float attribution_height = 8 + painter.font().pixel_size() + 8;
+        if (
+            static_cast<float>(event.x()) > static_cast<float>(width()) - attribution_width && static_cast<float>(event.y()) > static_cast<float>(height()) - attribution_height) {
+            Desktop::Launcher::open(m_attribution_url);
+            return;
+        }
+    }
+
+    // Start map tiles dragging
+    m_dragging = true;
+    m_last_mouse_x = event.x();
+    m_last_mouse_y = event.y();
+    set_override_cursor(Gfx::StandardCursor::Drag);
+}
+
+void MapsWidget::mousemove_event(GUI::MouseEvent& event)
+{
+    if (!m_dragging)
+        return;
+
+    double delta_x = event.x() - m_last_mouse_x;
+    double delta_y = event.y() - m_last_mouse_y;
+
+    // Adjust map center latitude and longitude by mouse delta
+    double tile_size = TILE_SIZE * pow(2, m_zoom);
+    m_center.latitude -= (-delta_y / tile_size) * 180;
+    m_center.latitude = min(max(m_center.latitude, -MAX_LATITUDE), MAX_LATITUDE);
+    m_center.longitude -= (delta_x / tile_size) * 360;
+    m_center.longitude = min(max(m_center.longitude, -180), 180);
+
+    m_last_mouse_x = event.x();
+    m_last_mouse_y = event.y();
+}
+
+void MapsWidget::mouseup_event(GUI::MouseEvent&)
+{
+    // Stop map tiles dragging
+    if (m_dragging) {
+        m_dragging = false;
+        set_override_cursor(Gfx::StandardCursor::Arrow);
+    }
+}
+
+void MapsWidget::mousewheel_event(GUI::MouseEvent& event)
+{
+    // Handle zooming in and out
+    if (event.wheel_delta_y() > 0) {
+        m_zoom--;
+    } else {
+        m_zoom++;
+    }
+    m_zoom = min(max(m_zoom, 2), 19);
+}
+
+MapsWidget::Tile* MapsWidget::get_tile(int x, int y)
+{
+    // Get the right tile from tiles cache
+    for (auto* tiles : m_tiles)
+        if (tiles->x == x && tiles->y == y && tiles->zoom == m_zoom)
+            return tiles;
+
+    // Load tile image from internet when not in cache
+    Tile* tile = new Tile { x, y, m_zoom };
+    if (m_tiles.size() > MAX_TILES)
+        m_tiles.remove(0);
+    m_tiles.append(tile);
+
+    // Create HTTP GET request to load image
+    HashMap<DeprecatedString, DeprecatedString> headers;
+    headers.set("User-Agent", "SerenityOS Maps");
+    headers.set("Accept", "*/*");
+    URL url(String::formatted(m_tile_layer_url, m_zoom, x, y).release_value_but_fixme_should_propagate_errors());
+    auto request = m_request_client->start_request("GET", url, headers, {});
+
+    // Ref it here so it wont go out of scope
+    request->ref();
```

Comment:
This feels sketchy. Maybe the MapsWidget could store the active requests instead?